### ISSUE TITLE
Question about Advantage Calculation and Prompt Grouping

### ISSUE THREAD

**Issue by ghyggg** [yifan123/flow_grpo#44]
Dear Author,

I’m currently working on reproducing your code and would like to ask a quick question. In the following line where advantages are computed, it seems that the samples do not come from the same prompt, correct? I noticed that before this line, there is a shuffling operation using perm applied to the samples.

perm = torch.randperm(total_batch_size, device=accelerator.device)
            # perm = torch.arange(total_batch_size, device=accelerator.device)
samples = {k: v[perm] for k, v in samples.items()}

advantages = torch.clamp(
                            sample["advantages"][:, j],
                            -config.train.adv_clip_max,
                            config.train.adv_clip_max,
                        )

**Comment by yifan123** [comment]
Here, the samples are pre-collected data, similar to a prepared training set in supervised learning. We shuffle them along the batch dimension using the same order for each key to maintain alignment.
